File: eval-reviews/2026-09-12-chained/query-null/typescript/astra/r1/starter/benchmark_views.ts
/** Local scaling probe, including initialization; no acceptance threshold.
 * Run: node starter/benchmark_views.ts */
import assert from "node:assert/strict";
import { performance } from "node:perf_hooks";
import { commands } from "./views.ts";

for (const size of [1000, 10000]) {
  const database = ["a", "b"].map(name => ({ name,
    columns: ["k", "v"].map(name => ({ name, type: "int", nullable: false })),
    rows: Array.from({ length: size }, (_, k) => [k, 1]),
  }));
  const sqls = [
    "SELECT COUNT(*) AS n, SUM(v) AS s FROM a",
    "SELECT COUNT(*) AS n, SUM(a.v) AS s FROM a LEFT JOIN b ON a.k = b.k",
    "SELECT SUM(v) AS s FROM b",
  ];
  const create = sqls.flatMap((sql, i) => [false, true].map((optimize, j) => ({ op: "create", view: `v${i}-${j}`, sql, optimize })));
  const updates = Array.from({ length: 800 }, (_, i) => [
    { op: "apply", changes: [{ op: "update", table: "a", id: 1, row: [0, i % 2 + 1] }] },
    { op: "read", view: "v1-1" },
  ]).flat();
  const durations = [];
  for (let trial = 0; trial < 5; trial++) {
    const start = performance.now();
    const result = commands({ protocol_version: 1, task: "query-null", input: { database, commands: [...create, ...updates] } });
    durations.push(performance.now() - start);
    assert.deepEqual(result.at(-1), { ok: true, result: { revision: 800, columns: ["n", "s"], rows: [[size, size + 1]] } });
  }
  durations.sort((a, b) => a - b);
  console.log(`${size * 2} base rows, six views, 800 applies + reads: ${durations[2].toFixed(1)} ms median including creation`);
}

File: eval-reviews/2026-09-12-chained/query-null/python/astra/r1/starter/benchmark_views.py
"""Local scaling smoke workload; no calibrated acceptance threshold."""
import time
from test_extension import table
from views import commands
for n in (2000, 8000):
    db = {name: table(name, [('k', 'int', False), ('v', 'int', True)], [[i, i] for i in range(n)]) for name in ('aa', 'bb')}
    cmds = [
        {'op': 'create', 'view': 'joined', 'sql': 'SELECT aa.k AS k, SUM(bb.v) AS s FROM aa LEFT JOIN bb ON aa.k = bb.k GROUP BY aa.k ORDER BY s DESC LIMIT 10', 'optimize': True},
        {'op': 'create', 'view': 'global', 'sql': 'SELECT COUNT(v) AS n, SUM(v) AS s FROM aa', 'optimize': False},
        {'op': 'create', 'view': 'unchanged', 'sql': 'SELECT COUNT(*) AS n FROM bb', 'optimize': True},
    ]
    for i in range(300):
        rid = i % n + 1
        cmds.append({'op': 'apply', 'changes': [{'op': 'update', 'table': 'aa', 'id': rid, 'row': [rid - 1, -i]}]})
        cmds.append({'op': 'read', 'view': 'global'})
    start = time.perf_counter()
    replies = commands(db, cmds)
    elapsed = time.perf_counter() - start
    assert all(r['ok'] for r in replies)
    assert replies[-1]['result']['rows'] == [[n, n*(n-1)//2 - 2*sum(range(300))]]
    print(f'{2*n} initial rows, 3 views, 300 updates/reads: {elapsed:.3f}s')
